src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,124 @@
+package htsjdk.samtools.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncWriterPool<A> implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed;
+    private final ArrayList<PooledWriter<A>> writers;
+
+    public AsyncWriterPool(final int threads) {
+        assert threads >= 1; // TODO: throw error
+
+        this.poolClosed = new AtomicBoolean(false);
+        // TODO: make this structure threadsafe
+        this.writers = new ArrayList<>();
+
+        executor = Executors.newWorkStealingPool(threads);
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (PooledWriter<A> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+    public static class PooledWriter<B> implements WrappedWriter<B> {
+        private final WrappedWriter<B> writer;
+        private final BlockingQueue<B> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+        public final AsyncWriterPool<B> outer;
```

Comment:
You shouldn't need this.  Instead make the class be `public class PooledWriter` without the `static` keyword.  Then you can access the outer with `AsyncWriterPool.this`.